### PR TITLE
Minimal warnings from GCC/Clang with strict conformance compilation; improvement in MSVC strict conformance compilation; fixed const qualifier issue with fvar type resolution

### PR DESCRIPTION
**Pull request by kedarbhat**
Contains all changes included in PR #14.

Changes between this PR and PR #14:
GCC/Clang: Compiling with flags `-v -ftemplate-backtrace-limit=0 -Weverything -Wall -Wextra -Wpedantic -Wno-c++98-compat -Wno-c++98-compat-pedantic -Wno-used-but-marked-unused -Wno-global-constructors -Wno-disabled-macro-expansion -Wno-double-promotion -ggdb`, we have minimal warnings (-Wpadding and -Wheader-hygiene remain)

MSVC: Compiling with switches `/W4 /permissive- /bigobj /EHsc`, most Autodiff sourced issues are gone, but some remain, particularly for MSVC 17. 

I was able to add a fix for issue #15 using std::decay and type aliases without having to change much of the code. This is part of this PR.

### CODE REVIEW

File: include/boost/math/differentiation/autodiff.hpp [hunk 1]
```diff
@@ -9,8 +9,14 @@
 #include <boost/config.hpp>
 #include <boost/math/tools/config.hpp>
 
+#if defined(_MSC_VER) || defined(BOOST_MSVC)
+#define NOMINMAX
```
Comment (shaneasd) [inline include/boost/math/differentiation/autodiff.hpp]: #define in a header file is not ideal. Is this to overcome the ridiculous #defines of min and max in the windows headers? If so could the same be achieved by enclosing our calls to min/max in parentheses? i.e. ```auto m = (std::max)(a, b)```. I think I saw that pattern used elsewhere in this file (or at least this repo).

**Comment by kedarbhat** [comment]
These files are in beta stage at best. Before time of submission, necessary but messy code will move, and the rest will be removed.

That #define is related to Windows max/min, for which I’m using parens where needed.

File: include/boost/math/differentiation/autodiff.hpp (hunk 1 above)
Comment (kedarbhat) [reply include/boost/math/differentiation/autodiff.hpp]: These files are in beta stage at best. Before time of submission, necessary but messy code will move, and the rest will be removed.

That #define is related to Windows max/min, for which I’m using parens where needed.

Comment (kedarbhat) [reply include/boost/math/differentiation/autodiff.hpp]: I removed the code you highlighted.

**Comment by pulver** [comment]
IMO appeasing Microsoft warnings is a low priority and not worth adding complexity to autodiff.hpp for.

If you are currently walking through the dark valley of complexity searching for the light of simplicity, then please let me know when you find it. :)

**Comment by kedarbhat** [comment]
> IMO appeasing Microsoft warnings is a low priority and not worth adding complexity to autodiff.hpp for.
> 
> If you are currently walking through the dark valley of complexity searching for the light of simplicity, then please let me know when you find it. :)

This should be ready to merge. The Boost guidelines say a bit about being able to compile without warnings with strict conformance, so I’m keeping an eye on complexity metrics as I make changes. MSVC likely just needs conversion operators at this point, but I’m stepping through the breaks to see how little intrusion I can make to achieve the outcome.

I’ll likely try to replace the int conversion operator with a generic conversion operator so I don’t need to write an operator for each type that happens to need be static casted to a given arithmetic type.

**Comment by kedarbhat** [comment]
I'm closing this PR and will resubmit later this week (with more needed changes), we're giving Boost-CI priority (and since it's much less intensive than strict conformance fixes for all Clang/GCC/MSVC). I've moved the conformance changes to branch [kedarbhat/autodiff branch `feature/strict_conformance`](https://github.com/kedarbhat/autodiff/tree/feature/strict_conformance). 

I've lined [kedarbhat/autodiff branch `develop`](https://github.com/kedarbhat/autodiff/tree/develop) with [pulver/autodiff branch `develop`](https://github.com/pulver/autodiff/tree/develop) so that once Boost-CI integration is complete, a PR will be more easily understood.

**Pull request closed without merging**